<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="content-type" content="text/html; charset=utf-8">

    <!-- Enable responsiveness on mobile devices-->
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

    <title>Yet Another Bioinformatic blog</title>

    <!-- css -->
    <link rel="stylesheet" href="/cayman.css">
    <link rel="stylesheet" href="/add.css">

    <!-- font -->
    <link rel="stylesheet" href="/fonts/fork_awesome/css/fork-awesome.min.css">
    <link rel="stylesheet" href="/fonts/academicons/css/academicons.min.css">
    
    <!-- js -->

    

    
<script>
  MathJax = {
      tex: {
	  inlineMath: [['$', '$'], ['\\(', '\\)']]
      },
      svg: {
	  fontCache: 'global'
      }
  };
</script>

<script type="text/javascript" id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js"> </script>



    <!-- Fathom - simple website analytics - https://github.com/usefathom/fathom -->
<script>
(function(f, a, t, h, o, m){
	a[h]=a[h]||function(){
		(a[h].q=a[h].q||[]).push(arguments)
	};
	o=f.createElement('script'),
	m=f.getElementsByTagName('script')[0];
	o.async=1; o.src=t; o.id='fathom-script';
	m.parentNode.insertBefore(o,m)
})(document, window, '//stats.pierre.marijon.fr/tracker.js', 'fathom');
fathom('set', 'siteId', 'NYKUR');
fathom('trackPageview');
</script>
<!-- / Fathom -->



    
    
  </head>

  <body>

    <a href="https://github.com/natir/blog" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#485F41; color:#B6C8B1; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>

    
    <section class="page-header">
      <h1 class="project-name">Yet Another Bioinformatic blog</h1>
      <a class="btn" href="/">Home</a>
      <a class="btn" href="//pierre.marijon.fr">About</a>
      <a class="btn" href="/rss.xml">RSS feed</a>
    </section>


    <section class="main-content">
      
      <div class="posts">
        
	
        <section>
          <a class='header-link' href="https:&#x2F;&#x2F;blog.pierre.marijon.fr&#x2F;misassemblies-in-noisy-assemblies&#x2F;">
            <h3 class="title"> Misassemblies in noisy assemblies </h3>
          </a>
          <p> 
            I think that all the people who have ever done a genome assembly one day say: &amp;quot;Ok my assembly is cool, but now how I can be sure that it&#x27;s the best and it doesn&#x27;t contain a lot of errors ?&amp;quot;
We have many technics to evaluate the quality of assemb… 
          </p>

          <!-- TODO: modularize this so we don't repeat ourselves -->
          <div class='info'>
            <span>14 minute read</span>
            
            <span class='divider'/>
            <span>19 September 2019</span>
            
            
          </div>
        </section>
        <hr/>
	
        
	
        <section>
          <a class='header-link' href="https:&#x2F;&#x2F;blog.pierre.marijon.fr&#x2F;why-i-stopped-c&#x2F;">
            <h3 class="title"> Why I stopped C++ </h3>
          </a>
          <p> 
            I have the feeling that generally bioinformatician use two languages, a language for small scripts, rapid analysis, prototyping (usually an interpreted language) and another when we have performance needs (usually a compiled language).
Until recently thes… 
          </p>

          <!-- TODO: modularize this so we don't repeat ourselves -->
          <div class='info'>
            <span>6 minute read</span>
            
            <span class='divider'/>
            <span>12 November 2018</span>
            
            
          </div>
        </section>
        <hr/>
	
        
	
        <section>
          <a class='header-link' href="https:&#x2F;&#x2F;blog.pierre.marijon.fr&#x2F;binary-mapping-format&#x2F;">
            <h3 class="title"> How to reduce the impact of your PAF file on your disk by 95% </h3>
          </a>
          <p> 
            Last week Shaun Jackman posted this tweet:


I have a 1.2 TB PAF.gz file of minimap2 all-vs-all alignments of 18 flowcells of Oxford Nanopore reads. Yipes. I believe that&amp;#39;s my first file to exceed a terabyte. Is there a better way? Perhaps removing th… 
          </p>

          <!-- TODO: modularize this so we don't repeat ourselves -->
          <div class='info'>
            <span>10 minute read</span>
            
            <span class='divider'/>
            <span>11 October 2018</span>
            
            
          </div>
        </section>
        <hr/>
	
        
	
        <section>
          <a class='header-link' href="https:&#x2F;&#x2F;blog.pierre.marijon.fr&#x2F;long-reads-overlapper-compare&#x2F;">
            <h3 class="title"> State-of-the-art long reads overlapper-compare </h3>
          </a>
          <p> 
            Introduction
In 2017, Chu et la. wrote a review 1 to present and compare 5 long-read overlapping tools, on 4 datasets (including 2 synthetic ones).
This paper is very cool and clear. The authors compare overlappers with respect to peak memory, wall clock … 
          </p>

          <!-- TODO: modularize this so we don't repeat ourselves -->
          <div class='info'>
            <span>7 minute read</span>
            
            <span class='divider'/>
            <span>13 April 2018</span>
            
            
          </div>
        </section>
        <hr/>
	
        
      </div>

      
          
      <footer class="site-footer">
	<span class="site-footer-credits">
	  <center>
	    <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
	      <img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" />
	    </a>
	  </center>
	  <br />
	  <span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">Yet Another Bioinformatic blog</span> by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Pierre Marijon</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href=https:&#x2F;&#x2F;blog.pierre.marijon.fr&#x2F; rel="dct:source">blog.pierre.marijon.fr</a>.</span>
      </footer>
    </section>
    

    
    

  </body>
</html>
